store/src/main/java/org/apache/rocketmq/store/schedule/ScheduleMessageService.java

```diff
@@ -431,35 +492,414 @@ public void executeOnTimeup() {
+                            log.error("[BUG] the real topic of schedule msg is {}, discard the msg. msg={}",
+                                    msgInner.getTopic(), msgInner);
+                            continue;
+                        }
+
+                        // Asynchronous send flow control.
+                        int maxPendingLimit = ScheduleMessageService.this.defaultMessageStore
+                                .getMessageStoreConfig().getScheduleAsyncDeliverMaxPendingLimit();
+                        int currentPendingNum = this.getDeliverPendingQueue().size();
+                        if (currentPendingNum > maxPendingLimit) {
+                            log.warn("Asynchronous sending triggers flow control, " +
+                                    "currentPendingNum={}, maxPendingLimit={}", currentPendingNum, maxPendingLimit);
+                            this.toNextTask(currentOffset);
+                            return;
+                        }
+
+                        // Handling abnormal blocking.
+                        PutResultProcess firstProcess = this.getDeliverPendingQueue().peek();
+                        if (firstProcess != null && firstProcess.need2Wait()) {
+                            log.warn("Asynchronous sending block. info={}", firstProcess.toString());
+                            this.toNextTask(currentOffset);
+                            return;
+                        }
+
+                        CompletableFuture<PutMessageResult> future =
+                                ScheduleMessageService.this.writeMessageStore.asyncPutMessage(msgInner);
+                        PutResultProcess putResultProcess = new PutResultProcess()
+                                .setTopic(msgInner.getTopic())
+                                .setDelayLevel(this.delayLevel)
+                                .setOffset(currentOffset)
+                                .setPhysicOffset(offsetPy)
+                                .setPhysicSize(sizePy)
+                                .setMsgId(msgExt.getMsgId())
+                                .setAutoResend(true)
+                                .setFuture(future)
+                                .thenProcess();
+                        this.getDeliverPendingQueue().add(putResultProcess);
+                        continue;
+                    } catch (Exception e) {
+                        log.error("ScheduleMessageService, messageTimeup execute error, drop it. msgExt={}, nextOffset={}, offsetPy={}, sizePy={}", msgExt, currentOffset, offsetPy, sizePy, e);
+                    }
+                }
+
+                long nextOffset = offset + (i / ConsumeQueue.CQ_STORE_UNIT_SIZE);
+                this.toNextTask(nextOffset);
+                return;
+            } finally {
+                bufferCQ.release();
+            }
+        }
+
+        public void toNextTask(long offset) {
```

Comment:
method name `toNextTask` is ambiguous, `scheduleNextTimerTask` could be better

Reply:
Agree with your suggestion, thanks